--- a/src/cmd/ksh93/sh/xec.c    Tue Jul 16 12:43:53 2013
+++ b/src/cmd/ksh93/sh/xec.c    Tue Jul 16 12:44:07 2013
@@ -1572,7 +1572,7 @@
unset_instance(nq,&amp;node,&amp;nr,mode);
sh_funstaks(slp-&gt;slchild,-1);
                                         stkclose(slp-&gt;slptr);
-                                       if(jmpval &gt; SH_JMPFUN)
+                                       if(jmpval &gt; SH_JMPFUN || (io &amp;&amp; 
jmpval &gt; SH_JMPIO) )
siglongjmp(*shp-&gt;jmplist,jmpval);
                                         goto setexit;
                                 }
